### PR TITLE
add CustomGlobalExceptionHandler class and custom validator

### PR DESCRIPTION
**Pull request by nikitaabo**

### CODE REVIEW

File: src/main/java/com/example/spring/controller/BookController.java [hunk 1]
```diff
@@ -34,7 +35,7 @@ public BookDto getBookById(@PathVariable Long id) {
     }
 
     @PostMapping
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/BookController.java:41]: Add `@Valid` to updateBook endpoint

Also, you can validate Long id by using `@Positive`. In such case add `@Validated` under the controller

File: src/main/java/com/example/spring/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
 
+import com.example.spring.validator.UniqueIsbn;
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
 import java.math.BigDecimal;
 import lombok.Data;
 
 @Data
 public class CreateBookRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

for String properties

File: src/main/java/com/example/spring/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
     private String title;
+    @NotNull
     private String author;
+    @NotNull
+    @UniqueIsbn
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/dto/CreateBookRequestDto.java]:
```suggestion
    @Isbn
```

File: src/main/java/com/example/spring/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+        body.put("errors", errors);
+        return new ResponseEntity<>(body, headers, status);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return new ResponseEntity<>(body, headers, HttpStatus.BAD_REQUEST);
```

File: src/main/java/com/example/spring/validator/UniqueIsbnValidator.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    private final BookRepository bookRepository;
+
+    @Override
+    public void initialize(UniqueIsbn constraintAnnotation) {
+    }
+
+    @Override
+    public boolean isValid(String isbn, ConstraintValidatorContext context) {
+        return isbn != null && !bookRepository.existsByIsbn(isbn);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/validator/UniqueIsbnValidator.java]: if you would like to validate isbn, it's better to use regexp. 
You should already have unique constraint for isbn in model so it's redundant validation
```
@Column(unique = true, nullable = false)
private String isbn;
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/spring/dto/CreateBookRequestDto.java [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
 
 @Data
 public class CreateBookRequestDto {
+    @NotBlank
     private String title;
+    @NotBlank
     private String author;
+    @NotBlank
     private String isbn;
+    @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/dto/CreateBookRequestDto.java]:
```suggestion
    @NotNull
```

it's BigDecimal

File: src/main/java/com/example/spring/repositories/book/BookRepository.java [hunk 7]
```diff
@@ -5,5 +5,5 @@
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
-
+    boolean existsByIsbn(String isbn);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/repositories/book/BookRepository.java]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/spring/dto/CreateBookRequestDto.java (hunk 6 above)
Comment (nikitaabo) [reply src/main/java/com/example/spring/dto/CreateBookRequestDto.java]: oh, yeah, thanks for note

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**